Migrate from Deferred to Promises, rework the imageCache

swederik opened the pull request:
@diego0020 Any thoughts?

diego0020 commented:
This looks great.

What is the purpose of *CornerstoneImageCacheFull* , I notice it's being triggered after some images have been removed from the cache so that it is no longer full. Perhaps it should be called before attempting to free space to let users use a different strategy to free space?
https://github.com/chafey/cornerstone/blob/1353e5100ab0221ae91fceb2f001759a5776d444/src/imageCache.js#L27

What happens if the promise is rejected in *putImageLoadObject* ? there should be a `catch(...)` which should remove the corresponding object from the *imageCacheDict*
https://github.com/chafey/cornerstone/blob/1353e5100ab0221ae91fceb2f001759a5776d444/src/imageCache.js#L96

In *removeImageLoadObject* the *cancelFn* should probably be called only if the image is not yet loaded  `if (cachedImage.loaded === false)` 
https://github.com/chafey/cornerstone/blob/1353e5100ab0221ae91fceb2f001759a5776d444/src/imageCache.js#L155

Promise rejection in *loadImageFromImageLoader* should be handled, probably triggering other event.
https://github.com/chafey/cornerstone/blob/1353e5100ab0221ae91fceb2f001759a5776d444/src/imageLoader.js#L42

Finally, it would be great to have some tests of cancelling an image. If you want I can help write some tests during the weekend.

diego0020 commented:
Another potential issue is calling the *cancelFn* from *removeImageLoadObject* from *purgeCacheIfNecessary*. 

For example consider the case when you are loading an image from a very slow source. The timestamp gets set as soon as you start loading. In the meantime the cache fills with other images while you are still waiting. The *purgeCacheIfNecessary* will be activated, which will in turn call the *cancelFn()* from the oldest image (which is still loading). This will result in a very confusing error for the user, and the image will fail to load.

I know it is a bizarre scenario, but maybe the *cancelFn* should not be called in this scenario. Maybe it should only be called explicitly by the user, or when purging cache.

swederik commented:
Thanks @diego0020 for the thorough review!

- We have been using CornerstoneImageCacheFull to pause prefetching when the image cache is full. This approach has actually been removed recently, so now I don't think we are using it for anything. I see it here: https://github.com/OHIF/Viewers/blob/fc29b48b591c0259453ba97da43f0394db7feaa7/Packages/ohif-viewerbase/client/lib/classes/StudyPrefetcher.js#L17
- I removed cancelFn usage from removeImageLoadObject. I think you're right. We could put it in purgeCache but it might not be a good idea to keep in removeImageLoadObject. We don't want to cancel requests unintentionally.
- I added rejection handling to imageLoader and putImageLoadObject.

Of course I would be happy for any help with tests!
